### PR TITLE
DAOS-11400 bio: Meta & WAL layout

### PR DESCRIPTION
**Pull request by NiuYawei**
Defined Meta & WAL blob header, WAL durable formats, Opaque bio_mc_context, BIO meta context management APIs.

Required-githooks: true

Signed-off-by: Niu Yawei <yawei.niu@intel.com>

### CODE REVIEW

File: src/bio/bio_wal.h [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+	uint64_t	mh_padding3;		/* Reserved */
+	uint32_t	mh_padding4;		/* Reserved */
+	uint16_t	mh_csum_type;
+	uint16_t	mh_csum_len;
+	uint8_t		mh_csum[0];		/* Checksum of this header */
+};
+
+/* WAL blob header */
+struct wal_header {
+	uint32_t	wh_magic;
```
Comment (wangshilong) [inline src/bio/bio_wal.h:37]: could we add a wh_version here too, in case wal_header need be extended in the future.

File: src/bio/bio_wal.h [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+}
+
+enum wal_hdr_flags {
+	WAL_HDR_FL_TAIL	= (1UL << 0),	/* The tail entry is in current block */
+};
+
+/* WAL transaction header */
+struct wal_trans_head {
+	uint32_t	th_magic;
+	uint32_t	th_len;		/* Total length of transaction entries in WAL */
```
Comment (wangshilong) [inline src/bio/bio_wal.h:59]: maybe a th_version here too..

File: src/bio/bio_wal.h [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+	uuid_t		mh_wal_devid;		/* WAL SSD device ID */
+	uuid_t		mh_data_devid;		/* Data SSD device ID */
+	uint64_t	mh_meta_blobid;		/* Meta blob ID */
+	uint64_t	mh_wal_blobid;		/* WAL blob ID */
+	uint64_t	mh_data_blobid;		/* Data blob ID */
+	uint32_t	mh_blk_bytes;		/* Block size for meta, in bytes */
+	uint32_t	mh_hdr_blks;		/* Meta blob header size, in blocks */
+	uint64_t	mh_tot_blks;		/* Meta blob capacity, in blocks */
+	uint32_t	mh_vos_id;		/* Associated per-engine target ID */
+	uint32_t	mh_padding1;		/* Reserved */
```
Comment (wangshilong) [inline src/bio/bio_wal.h:26]: this could be simplified as uint32_t mp_paddings[6];

**Review by wangshilong — APPROVED** [review]

**Review by wangshilong — DISMISSED** [review]

File: src/bio/bio_wal.h [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+	uint64_t	wh_tot_blks;	/* WAL blob capacity, in blocks */
+	uint64_t	wh_ckp_id;	/* Last check-pointed transaction ID */
+	uint64_t	wh_commit_id;	/* Last committed transaction ID */
+	uint64_t	wh_next_id;	/* Next unused transaction ID */
+	uint64_t	wh_padding1;	/* Reserved */
+	uint32_t	wh_padding2;	/* Reserved */
+	uint16_t	wh_padding3;	/* Reserved */
+	uint16_t	wh_csum_len;
+	uint8_t		wh_csum[0];	/* Checksum of this header */
+}
```
Comment (wangshilong) [inline src/bio/bio_wal.h:48]: [defect] missed ';'

File: src/include/daos_srv/bio.h [hunk 5]
```diff
@@ -909,4 +909,60 @@ int bio_copy(struct bio_io_context *ioctxt, struct bio_sglist *bsgl_src,
...
+ * Open Meta/Data/WAL blobs, load WAL header, alloc meta context
+ *
+ * \param[in]	xs_ctxt		Per-xstream NVMe context
+ * \param[in]	pool_id		Pool UUID
+ * \param[in]	umm		umem instance
+ * \param[out]	mc		BIO meta context
+ *
+ * \return			Zero on success, negative value on error
+ */
+int bio_mc_open(struct bio_xs_context *xs_ctxt, uuid_t pool_id, struct umem_instance *umm,
```
Comment (wangshilong) [inline src/include/daos_srv/bio.h:956]: maybe we need a skip_blob here to indicate there is no nvme usage, but we still need return a bio_meta_context to issue io to PMEM only.

see bio_ioctxt_open()

File: src/bio/bio_wal.h [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+	uint8_t		tt_csum[0];	/* Checksum of WAL transaction */
+};
+
+/* Meta context */
+struct bio_meta_context {
+	struct bio_io_context	*mc_data;	/* Data blob I/O context */
+	struct bio_io_context	*mc_meta;	/* Meta blob I/O context */
+	struct bio_io_context	*mc_wal;	/* WAL blob I/O context */
+	struct wal_header	 mc_wal_header;	/* WAL header */
+	struct ABT_cond		 mc_commit_wq;	/* FIFO waitqueue for trans commit */
```
Comment (wangshilong) [inline src/bio/bio_wal.h:94]: Just ABT_cond is fine...
src/bio/bio_wal.h:94:20: error: field ‘mc_commit_wq’ has incomplete type
  struct ABT_cond   mc_commit_wq; /* FIFO waitqueue for trans commit */
                    ^~~~~~~~~~~~
src/bio/bio_wal.h:95:20: error: field ‘mc_begin_wq’ has incomplete type
  struct ABT_cond   mc_begin_wq; /* FIFO waitqueue for trans begin */

File: src/bio/bio_wal.h (hunk 4 above)
Comment (NiuYawei) [reply src/bio/bio_wal.h:48]: sure, thanks.

File: src/bio/bio_wal.h (hunk 6 above)
Comment (NiuYawei) [reply src/bio/bio_wal.h:94]: oops, thanks.

File: src/include/daos_srv/bio.h (hunk 5 above)
Comment (NiuYawei) [reply src/include/daos_srv/bio.h:956]: Ok.

**Review by wangshilong — DISMISSED** [review]

File: src/bio/bio_wal.h [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+	uint64_t	mh_wal_blobid;		/* WAL blob ID */
+	uint64_t	mh_data_blobid;		/* Data blob ID */
+	uint32_t	mh_blk_bytes;		/* Block size for meta, in bytes */
+	uint32_t	mh_hdr_blks;		/* Meta blob header size, in blocks */
+	uint64_t	mh_tot_blks;		/* Meta blob capacity, in blocks */
+	uint32_t	mh_vos_id;		/* Associated per-engine target ID */
+	uint32_t	mh_padding[6];		/* Reserved */
+	uint16_t	mh_csum_type;
+	uint16_t	mh_csum_len;
+	uint8_t		mh_csum[0];		/* Checksum of this header */
```
Comment (jolivier23) [inline src/bio/bio_wal.h:29]: do we need to use same csum algorithm as the container or can we just assume something simple and fixed size like crc32 ?  I presume we will have checksum regardless ?

**Review by jolivier23 — APPROVED** [review]

File: src/bio/bio_wal.h (hunk 7 above)
Comment (NiuYawei) [reply src/bio/bio_wal.h:29]: Right, maybe it's not necessary to make the csum_type configurable on format time, crc32 is good enough, we can always use crc32.

**Review by wangshilong — DISMISSED** [review]
Minor styles could be addressed if pr need be refreshed to make CI happy.

**Review by wangshilong — APPROVED** [review]

**Review by jolivier23 — APPROVED** [review]

**Comment by NiuYawei** [comment]
@jolivier23 , are you the gatekeeper of this feature branch? I think this PR could be force landed, what do you think?

**Pull request merged**